packages/component-base/src/gestures.js

```diff
@@ -0,0 +1,1096 @@
+  },
+  /**
+   * @this {GestureRecognizer}
+   * @param {TouchEvent} e
+   * @return {void}
+   */
+  touchend: function (e) {
+    downupFire('up', _findOriginalTarget(e), e.changedTouches[0], e);
+  }
+});
+
+/**
+ * @param {string} type
+ * @param {EventTarget} target
+ * @param {Event|Touch} event
+ * @param {Event=} preventer
+ * @return {void}
+ */
+function downupFire(type, target, event, preventer) {
+  if (!target) {
+    return;
+  }
+  _fire(target, type, {
+    x: event.clientX,
+    y: event.clientY,
+    sourceEvent: event,
+    preventer: preventer,
+    prevent: function (e) {
+      return prevent(e);
+    }
+  });
+}
+
+register({
+  name: 'track',
+  touchAction: 'none',
+  deps: ['mousedown', 'touchstart', 'touchmove', 'touchend'],
+  flow: {
+    start: ['mousedown', 'touchstart'],
+    end: ['mouseup', 'touchend']
+  },
+  emits: ['track'],
+
+  info: {
+    x: 0,
+    y: 0,
+    state: 'start',
+    started: false,
+    moves: [],
+    /** @this {GestureInfo} */
+    addMove: function (move) {
+      if (this.moves.length > TRACK_LENGTH) {
+        this.moves.shift();
+      }
+      this.moves.push(move);
+    },
+    movefn: null,
+    upfn: null,
+    prevent: false
+  },
+
+  /**
+   * @this {GestureRecognizer}
+   * @return {void}
+   */
+  reset: function () {
+    this.info.state = 'start';
+    this.info.started = false;
+    this.info.moves = [];
+    this.info.x = 0;
+    this.info.y = 0;
+    this.info.prevent = false;
+    untrackDocument(this.info);
+  },
+
+  /**
+   * @this {GestureRecognizer}
+   * @param {MouseEvent} e
+   * @return {void}
+   */
+  mousedown: function (e) {
+    if (!hasLeftMouseButton(e)) {
+      return;
+    }
+    let t = _findOriginalTarget(e);
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    let self = this;
+    let movefn = function movefn(e) {
+      let x = e.clientX,
+        y = e.clientY;
+      if (trackHasMovedEnough(self.info, x, y)) {
+        // first move is 'start', subsequent moves are 'move', mouseup is 'end'
+        self.info.state = self.info.started ? (e.type === 'mouseup' ? 'end' : 'track') : 'start';
+        if (self.info.state === 'start') {
+          // if and only if tracking, always prevent tap
+          prevent('tap');
+        }
+        self.info.addMove({ x: x, y: y });
+        if (!hasLeftMouseButton(e)) {
+          // always fire "end"
+          self.info.state = 'end';
+          untrackDocument(self.info);
+        }
+        if (t) {
+          trackFire(self.info, t, e);
+        }
+        self.info.started = true;
+      }
+    };
+    let upfn = function upfn(e) {
+      if (self.info.started) {
+        movefn(e);
+      }
+
+      // remove the temporary listeners
+      untrackDocument(self.info);
+    };
+    // add temporary document listeners as mouse retargets
+    trackDocument(this.info, movefn, upfn);
+    this.info.x = e.clientX;
+    this.info.y = e.clientY;
+  },
```

Comment:
nit: let's add newlines between listener functions.

```suggestion
  },

```

Reply:
Done